### PR TITLE
Create LtiIntegration model

### PR DESCRIPTION
**Pull request by nicklathe**
This PR adds a new `LtiIntegration` model. I'm looking for feedback on any standards we have that I'm violating, or best practices I've missed. It includes:

* Model definition
* ~Migration~
* ~Schema change~
* Tests
* Updates to factories.rb

## Links

- spec: [LTI Tech Spec](https://docs.google.com/document/d/1tTwAF40cnn2oUvK2CmRO-a5qYXkCR5rR5gBJS2-zV_U/edit#heading=h.y0j9qhlwm35z)
- jira ticket: [PLT-20-101](https://codedotorg.atlassian.net/browse/P20-101)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/db/migrate/20230503220558_create_lti_integrations.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class CreateLtiIntegrations < ActiveRecord::Migration[6.0]
+  def change
+    create_table :lti_integrations do |t|
```
Comment (nicklathe) [inline dashboard/db/migrate/20230503220558_create_lti_integrations.rb]: Look into using [validations](https://guides.rubyonrails.org/active_record_validations.html) for some of these fields that should be required.

File: dashboard/test/models/lti_integration_test.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class LtiIntegrationTest < ActiveSupport::TestCase
```
Comment (nicklathe) [inline dashboard/test/models/lti_integration_test.rb:3]: Add some tests that test the validations.

File: dashboard/db/migrate/20230503220558_create_lti_integrations.rb (hunk 1 above)
Comment (nicklathe) [reply dashboard/db/migrate/20230503220558_create_lti_integrations.rb]: Required:
* platform_id
* issuer
* client_id
* platform_name
* auth_redirect_url
* jwks_url
* access_token_url

File: dashboard/test/factories/lti_integrations.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+FactoryBot.define do
```
Comment (nicklathe) [inline dashboard/test/factories/lti_integrations.rb]: Common practice here is to remove the generated factory file and add any factories to the shared `dashboard/test/factories/factories.rb`

**Comment by sureshc** [comment]
Because it’s hard to revert a Rails database/schema migration, we have an undocumented practice of putting the actual schema change (the migration and the new schema.rb) in its own Pull Request and merging that before any feature branches that implement validations, tests, load fixtures, etc. This reduces risks if the migration causes issues and needs to be rolled back somehow. Sorry this is not documented anywhere! We should write up some ActiveRecord best practices and/or lint rules to better guide the team on this.

**Comment by nicklathe** [comment]
> Because it’s hard to revert a Rails database/schema migration, we have an undocumented practice of putting the actual schema change (the migration and the new schema.rb) in its own Pull Request and merging that before any feature branches that implement validations, tests, load fixtures, etc. This reduces risks if the migration causes issues and needs to be rolled back somehow. Sorry this is not documented anywhere! We should write up some ActiveRecord best practices and/or lint rules to better guide the team on this.

Ok, this [documentation exists](https://docs.google.com/document/d/1l5ISiVT329Y9uKLZYPmrSvu63O3DrB2iVX4xKsfeQhA/edit#heading=h.r5fucihexafn). I must have misread it as this PR _should_ include the model changes and any other files generated from `rails generate migration...`. 

Will update PR to include only the `schema.rb` and the `20230503220558_create_lti_integrations.rb` files.

**Comment by nicklathe** [comment]
@sureshc here is a separate PR that adds only the migration and schema. Once that is approved and merged, I will rebase this PR to add the additional model, tests, and factory changes.

https://github.com/code-dot-org/code-dot-org/pull/51889

File: dashboard/test/models/lti_integration_test.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
```
Comment (cat5inthecradle) [inline dashboard/test/models/lti_integration_test.rb:3]: I am incredibly pro-test person, so this feedback isn't coming from a place of "we don't need no stinkin' tests!" :wink: In fact, this looks test-driven (:heart_eyes:) based on the order of your tests. However, they seem to be coming close to testing Rails vs testing our business logic. Are these tests useful?

The business logic here is: There are 7 required fields for an LtiIntegration object. Would it be cleaner to refactor these to specifically check validity in a single test case?

```ruby
test "should validate required fields" do
    assert_not build(:lti_integration, platform_id: nil).valid? "Missing platform_id should be invalid"
    assert_not build(:lti_integration, issuer: nil).valid? "Missing issuer should be invalid"
    # ...
end
```

My general nit here is that it looks like we're testing a lot, when really we aren't.

File: dashboard/test/factories/factories.rb [hunk 5]
```diff
@@ -1680,4 +1680,14 @@
...
+
+  factory :lti_integration do
+    platform_id {"1a2b3c5b"}
+    issuer {"http://some.lms.com"}
+    client_id {"12345678"}
+    platform_name {"canavs"}
+    auth_redirect_url {"http://some.lms.com/lti/authenticate"}
+    jwks_url {"http://some.lms.com/jwks"}
+    access_token_url {"http://some.lms.com/access_token"}
+  end
```
Comment (cat5inthecradle) [inline dashboard/test/factories/factories.rb]: Do we tend to use hardcoded values in factories, or do we ever prefer randomization? I honestly don't know. I'm a big fan of random so we discover edge cases earlier.

File: dashboard/test/models/lti_integration_test.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    integration = build :lti_integration, access_token_url: nil
+    assert_not integration.save, "Saved the LtiIntegration without access_token_url"
+  end
+
+  test "should save an LtiIntegration with required values" do
+    integration = build :lti_integration
+    assert integration.save, "Saved the LtiIntegration with required values"
+  end
+
+  test "should save an LtiIntegration with required values and name and admin_email" do
```
Comment (cat5inthecradle) [inline dashboard/test/models/lti_integration_test.rb]: Aren't `name` and `admin_email` NOT required?

File: dashboard/test/models/lti_integration_test.rb (hunk 4 above)
Comment (nicklathe) [reply dashboard/test/models/lti_integration_test.rb:3]: Thanks for the feedback. I agree, these tests should be slimmed down as they don't provide real value.

File: dashboard/test/factories/factories.rb (hunk 5 above)
Comment (nicklathe) [reply dashboard/test/factories/factories.rb]: Interesting, I'll take a look at some other factories for inspiration.

Comment (nicklathe) [reply dashboard/test/factories/factories.rb]: It seems like a lot of values in factories are hardcoded. However I've changed these to just be string values named as their key. If there is a preferred standard, I'm happy to follow it.

File: dashboard/test/models/lti_integration_test.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+require 'test_helper'
+
+class LtiIntegrationTest < ActiveSupport::TestCase
+  test "should validate required fields" do
+    assert_not build(:lti_integration, platform_id: nil).valid? "Missing platform_id should be invalid"
```
Comment (elf-code) [inline dashboard/test/models/lti_integration_test.rb]: Is the language here standard? I think "platform_id is required" is much more direct.

**Review by elf-code — APPROVED** [review]
LGTM, left one nit on test language.

File: dashboard/test/models/lti_integration_test.rb (hunk 7 above)
Comment (nicklathe) [reply dashboard/test/models/lti_integration_test.rb]: Agreed, I'll make change the language to be more direct.

**Review by cat5inthecradle — APPROVED** [review]

**Pull request merged**